File: crates/sweep-engine/src/guardrails.rs
//! Scan-path guardrails aligned with the TypeScript reference (`guardrails.ts`).

use std::collections::HashSet;
use std::path::{Component, Path, PathBuf};
use sweep_errors::{EngineError, GuardrailError};

const PROTECTED_VCS_DIR_NAMES: &[&str] = &[".git", ".svn", ".hg", ".bzr"];

/// Assert that the target directory is safe to operate on.
pub fn assert_safe_cwd(target_path: &str) -> Result<(), EngineError> {
    if target_path.contains('\0') {
        return Err(GuardrailError::ProtectedPath {
            path: target_path.to_owned(),
        }
        .into());
    }

    if target_path.contains("..") {
        return Err(GuardrailError::ProtectedPath {
            path: format!("Path traversal detected: {target_path}"),
        }
        .into());
    }

    let resolved = normalize_path(target_path);

    if blocked_roots().contains(&resolved) {
        return Err(GuardrailError::ProtectedPath {
            path: format!(
                "Refusing to operate on protected path: {}\n  \
                 sweep must be run inside a project directory, not at a system root.",
                resolved.display()
            ),
        }
        .into());
    }

    let root = filesystem_root(&resolved);
    let relative_parts = path_segments_below_root(&resolved, &root);
    if relative_parts.len() < 2 {
        return Err(GuardrailError::ProtectedPath {
            path: format!(
                "Path is too shallow to be a project directory: {}\n  \
                 Expected at least 2 path segments below filesystem root.",
                resolved.display()
            ),
        }
        .into());
    }

    Ok(())
}

/// Assert that a pattern string is safe (won't escape the target directory).
pub fn assert_safe_pattern(pattern: &str) -> Result<(), EngineError> {
    if pattern.trim().is_empty() {
        return Err(GuardrailError::ProtectedPath {
            path: "Pattern must not be empty.".to_owned(),
        }
        .into());
    }

    if pattern != pattern.trim() {
        return Err(GuardrailError::ProtectedPath {
            path: format!("Pattern must not have leading or trailing whitespace: \"{pattern}\""),
        }
        .into());
    }

    if pattern.contains('\0') {
        return Err(GuardrailError::ProtectedPath {
            path: format!("Pattern contains null byte: {pattern:?}"),
        }
        .into());
    }

    if pattern.starts_with('/') {
        return Err(GuardrailError::ProtectedPath {
            path: format!(
                "Patterns must not start with /: \"{pattern}\"\n  \
                 Use directory names or glob patterns like \"*.tsbuildinfo\"."
            ),
        }
        .into());
    }

    if pattern.contains("..") {
        return Err(GuardrailError::ProtectedPath {
            path: format!("Patterns must not contain \"..\" traversal: \"{pattern}\""),
        }
        .into());
    }

    Ok(())
}

/// Validate all pattern strings in a sweep config before scanning.
pub fn assert_safe_config_patterns(config: &sweep_types::SweepConfig) -> Result<(), EngineError> {
    for pattern in &config.patterns {
        assert_safe_pattern(pattern)?;
    }
    for pattern in &config.disabled_patterns {
        assert_safe_pattern(pattern)?;
    }
    for pattern in &config.ignore {
        assert_safe_pattern(pattern)?;
    }
    Ok(())
}

fn blocked_roots() -> HashSet<PathBuf> {
    let mut roots = HashSet::from([
        PathBuf::from("/"),
        PathBuf::from("/home"),
        PathBuf::from("/usr"),
        PathBuf::from("/usr/local"),
        PathBuf::from("/etc"),
        PathBuf::from("/opt"),
        PathBuf::from("/var"),
        PathBuf::from("/bin"),
        PathBuf::from("/sbin"),
        PathBuf::from("/lib"),
        PathBuf::from("/lib64"),
        PathBuf::from("/boot"),
        PathBuf::from("/sys"),
        PathBuf::from("/proc"),
        PathBuf::from("/dev"),
    ]);

    if let Some(home) = home_dir() {
        roots.insert(home);
    }

    #[cfg(windows)]
    {
        for drive in b'A'..=b'Z' {
            let letter = drive as char;
            let root = format!("{letter}:\\");
            roots.insert(PathBuf::from(&root));
            roots.insert(PathBuf::from(format!("{root}Windows")));
            roots.insert(PathBuf::from(format!("{root}Program Files")));
            roots.insert(PathBuf::from(format!("{root}Program Files (x86)")));
            roots.insert(PathBuf::from(format!("{root}Users")));
            roots.insert(PathBuf::from(format!("{root}ProgramData")));
        }
    }

    roots
}

fn home_dir() -> Option<PathBuf> {
    if let Ok(home) = std::env::var("HOME") {
        if !home.is_empty() {
            return Some(PathBuf::from(home));
        }
    }
    #[cfg(windows)]
    {
        if let Ok(userprofile) = std::env::var("USERPROFILE") {
            if !userprofile.is_empty() {
                return Some(PathBuf::from(userprofile));
            }
        }
    }
    None
}

fn normalize_path(path: &str) -> PathBuf {
    let path = Path::new(path);
    let mut normalized = PathBuf::new();
    for component in path.components() {
        match component {
            Component::Prefix(prefix) => normalized.push(prefix.as_os_str()),
            Component::RootDir => normalized.push(component.as_os_str()),
            Component::CurDir => {}
            Component::ParentDir => {
                normalized.pop();
            }
            Component::Normal(part) => normalized.push(part),
        }
    }
    normalized
}

fn filesystem_root(path: &Path) -> PathBuf {
    if cfg!(windows) {
        path.components()
            .next()
            .map(|component| PathBuf::from(component.as_os_str()))
            .unwrap_or_else(|| PathBuf::from("/"))
    } else {
        PathBuf::from("/")
    }
}

fn path_segments_below_root(resolved: &Path, parsed_root: &Path) -> Vec<String> {
    resolved
        .strip_prefix(parsed_root)
        .unwrap_or(resolved)
        .components()
        .filter_map(|component| match component {
            Component::Normal(part) => Some(part.to_string_lossy().into_owned()),
            _ => None,
        })
        .collect()
}

/// Whether any path segment is a protected VCS metadata directory.
pub fn path_has_protected_vcs_segment(entry_path: &str) -> bool {
    entry_path
        .split(['/', '\\'])
        .any(|segment| PROTECTED_VCS_DIR_NAMES.contains(&segment))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn assert_safe_cwd_rejects_root() {
        match assert_safe_cwd("/") {
            Err(EngineError::Guardrail(_)) => {}
            other => panic!("expected guardrail error, got {other:?}"),
        }
    }

    #[test]
    fn assert_safe_cwd_rejects_shallow_tmp() {
        match assert_safe_cwd("/tmp") {
            Err(EngineError::Guardrail(_)) => {}
            other => panic!("expected guardrail error, got {other:?}"),
        }
    }

    #[test]
    fn assert_safe_cwd_allows_project_path() {
        assert!(assert_safe_cwd("/tmp/user/project").is_ok());
    }

    #[test]
    fn assert_safe_pattern_rejects_absolute_paths() {
        match assert_safe_pattern("/etc") {
            Err(EngineError::Guardrail(_)) => {}
            other => panic!("expected guardrail error, got {other:?}"),
        }
    }
}

File: packages/core/src/guardrails.ts
import { lstatSync, realpathSync } from "node:fs";
import { homedir } from "node:os";
import { isAbsolute, normalize, parse, relative, resolve, sep } from "node:path";

// ─── Blocked paths ────────────────────────────────────────────────────────────

/** VCS dirs — never delete artifacts inside these path segments. */
export const PROTECTED_VCS_DIR_NAMES = new Set([".git", ".svn", ".hg", ".bzr"]);

/**
 * Paths that must never be the target directory.
 * Evaluated AFTER resolve() — these are canonical absolute paths.
 * Built once at module load (not per-call) for performance.
 */
function buildBlockedRoots(): Set<string> {
  const roots = new Set<string>([
    "/",
    "/home",
    "/usr",
    "/usr/local",
    "/etc",
    "/opt",
    "/var",
    "/bin",
    "/sbin",
    "/lib",
    "/lib64",
    "/boot",
    "/sys",
    "/proc",
    "/dev",
    homedir(),
  ]);

  if (process.platform === "win32") {
    for (const drive of "ABCDEFGHIJKLMNOPQRSTUVWXYZ") {
      const letter = `${drive}:\\`;
      roots.add(normalize(letter));
      roots.add(normalize(`${letter}Windows`));
      roots.add(normalize(`${letter}Program Files`));
      roots.add(normalize(`${letter}Program Files (x86)`));
      roots.add(normalize(`${letter}Users`));
      roots.add(normalize(`${letter}ProgramData`));
    }
  }

  return roots;
}

const BLOCKED_ROOTS = buildBlockedRoots();

// ─── Error type ───────────────────────────────────────────────────────────────

export class GuardrailError extends Error {
  /** Maps to process.exit() code */
  readonly code: number;

  constructor(message: string, code = 2) {
    super(message);
    this.name = "GuardrailError";
    this.code = code;
  }
}

// ─── Checks ───────────────────────────────────────────────────────────────────

/**
 * Assert that the target directory is safe to operate on.
 * Throws GuardrailError (exit code 2) if not.
 */
export function assertSafeCwd(targetPath: string): void {
  // Reject null bytes — can confuse C-level FS calls
  if (targetPath.includes("\x00")) {
    throw new GuardrailError(`Path contains null byte: ${JSON.stringify(targetPath)}`);
  }

  const resolved = normalize(resolve(targetPath));

  if (BLOCKED_ROOTS.has(resolved)) {
    throw new GuardrailError(
      `Refusing to operate on protected path: ${resolved}\n` +
        `  sweep must be run inside a project directory, not at a system root.`,
    );
  }

  // Must be at least 2 path segments deep (e.g., /home/user → ok, /tmp → blocked)
  const { root } = parse(resolved);
  const relativeParts = pathSegmentsBelowRoot(resolved, root);
  if (relativeParts.length < 2) {
    throw new GuardrailError(
      `Path is too shallow to be a project directory: ${resolved}\n` +
        `  Expected at least 2 path segments below filesystem root.`,
    );
  }
}

/**
 * Assert that a pattern string is safe (won't escape the target directory).
 */
export function assertSafePattern(pattern: string): void {
  if (!pattern || pattern.trim().length === 0) {
    throw new GuardrailError("Pattern must not be empty.");
  }
  if (pattern !== pattern.trim()) {
    throw new GuardrailError(`Pattern must not have leading or trailing whitespace: "${pattern}"`);
  }
  if (pattern.includes("\x00")) {
    throw new GuardrailError(`Pattern contains null byte: ${JSON.stringify(pattern)}`);
  }
  if (pattern.startsWith("/")) {
    throw new GuardrailError(
      `Patterns must not start with /: "${pattern}"\n` +
        `  Use directory names or glob patterns like "*.tsbuildinfo".`,
    );
  }
  if (pattern.includes("..")) {
    throw new GuardrailError(`Patterns must not contain ".." traversal: "${pattern}"`);
  }
}

/**
 * Assert that the estimated total size is within the configured limit.
 * Requires --force-large to bypass (which must be combined with --yes).
 */
export function assertSizeLimit(
  estimatedBytes: number,
  maxSizeGB: number,
  forceLarge: boolean,
): void {
  const estimatedGB = estimatedBytes / 1024 ** 3;
  if (estimatedGB > maxSizeGB && !forceLarge) {
    throw new GuardrailError(
      `Estimated size (${estimatedGB.toFixed(1)} GB) exceeds limit (${maxSizeGB} GB).\n` +
        `  Use --force-large --yes to proceed anyway.`,
    );
  }
}

/**
 * Check if a filesystem entry is a symlink.
 * Uses lstatSync to avoid following the link.
 */
export function isSymlink(entryPath: string): boolean {
  try {
    return lstatSync(entryPath).isSymbolicLink();
  } catch {
    return false;
  }
}

/**
 * True for symlinks and Windows junctions/reparse points that must not be recursed.
 */
export function isReparsePointOrSymlink(entryPath: string): boolean {
  try {
    const stat = lstatSync(entryPath);
    if (stat.isSymbolicLink()) {
      return true;
    }
    if (process.platform === "win32" && stat.isDirectory()) {
      try {
        const real = realpathSync.native(entryPath);
        return normalize(real) !== normalize(resolve(entryPath));
      } catch {
        return false;
      }
    }
  } catch {
    return false;
  }
  return false;
}

/** Path segments below the filesystem root (platform-aware). */
export function pathSegmentsBelowRoot(resolved: string, parsedRoot: string): string[] {
  const tail = resolved.slice(parsedRoot.length);
  return tail.split(sep).filter(Boolean);
}

/** Whether `candidatePath` is the target root or a path inside it. */
export function isPathWithinRoot(candidatePath: string, rootPath: string): boolean {
  const root = normalize(resolve(rootPath));
  const candidate = normalize(resolve(candidatePath));
  if (candidate === root) {
    return true;
  }
  if (process.platform === "win32" && candidate.toLowerCase() === root.toLowerCase()) {
    return true;
  }
  const rel = relative(root, candidate);
  if (rel === "") {
    return true;
  }
  if (rel.startsWith("..") || isAbsolute(rel)) {
    return false;
  }
  return true;
}

/**
 * Assert a candidate deletion path stays inside the plan target directory.
 * Throws GuardrailError (exit code 2) if not.
 */
export function assertPathWithinRoot(
  candidatePath: string,
  rootPath: string,
  label = "Candidate path",
): void {
  if (!isPathWithinRoot(candidatePath, rootPath)) {
    throw new GuardrailError(
      `${label} is outside the scan target:\n` +
        `  ${candidatePath}\n` +
        `  target: ${resolve(rootPath)}`,
    );
  }
}

/** True when any path segment is a protected VCS metadata directory. */
export function pathHasProtectedVcsSegment(entryPath: string): boolean {
  const segments = normalize(entryPath).split(sep);
  return segments.some((segment) => PROTECTED_VCS_DIR_NAMES.has(segment));
}
